arch/aarch64/mcount-dynamic.c

```diff
@@ -84,10 +87,21 @@ static void read_patchable_loc(struct mcount_dynamic_info *mdi, struct uftrace_e
 			       struct uftrace_elf_iter *iter, unsigned long offset)
 {
 	typeof(iter->shdr) *shdr = &iter->shdr;
+	unsigned i;
+	unsigned long *patchable_loc;
 
 	mdi->nr_patch_target = shdr->sh_size / sizeof(long);
 	mdi->patch_target = xmalloc(shdr->sh_size);
-
+	patchable_loc = mdi->patch_target;
+
+#ifdef __ANDROID__
+	/* symbol addresses on Android are stored in relocation records
+	   so read them from relocated section */
```

Comment:
On Android the relative offset is stored not in `__patchable_function_entries` but in `.rela.dyn` section.

Reply:
Firstly let me first clarify what's going on here. At runtime entries in `__patchable_function_entries` section contain absolute addresses of patched functions. Statically, in ELF file these entries may contain relative offsets (on Linux) or just zeroes (on Android). At startup, runtime linker will patch them by adding the load address of module.

Existing code reads static contents of `__patchable_function_entries` from ELF file but it's useless on Android as it's all zeroes. Instead I reconstruct relative function offsets by subtracting the module start address from already relocated contents of `__patchable_function_entries`.

I suspect that new code should also work on Linux but haven't tested it (`make test` does not seem to run tests for patched functions, only `-finstrument-functions` and `-pg`, perhaps that should be fixed?).

Reply:
> I suspect that new code should also work on Linux but haven't tested it

Ok, I tested it and it worked at least with s-abc.c. I'd be cautious to introduce this change on Linux without larger tests though.

BTW note that `patchable_gcc_nop` (and probly other patterns too) need to be updated because modern compilers prefix code with endbr64:
```
0000000000001169 <a>:
    1169:	f3 0f 1e fa          	endbr64 
    116d:	90                   	nop
    116e:	90                   	nop
    116f:	90                   	nop
    1170:	90                   	nop
    1171:	90                   	nop
    ...
```

Reply:
> Existing code reads static contents of __patchable_function_entries from ELF file but it's useless on Android as it's all zeroes.

Ok, good to know.  Yeah, if it contains absolute addresses at runtime we can read from the memory.  I was confused that you wanted to read from the `.rela.dyn` section.

Reply:
> `make test` does not seem to run tests for patched functions, only `-finstrument-functions` and `-pg`, perhaps that should be fixed?

That'd be nice :).

Reply:
> BTW note that patchable_gcc_nop (and probly other patterns too) need to be updated because modern compilers prefix code with endbr64.

Thanks for pointing that out.  We should update CET support.

Reply:
> Thanks for pointing that out. We should update CET support.

Added #1613

Reply:
> > `make test` does not seem to run tests for patched functions, only `-finstrument-functions` and `-pg`, perhaps that should be fixed?
> 
> That'd be nice :).

Added #1614

Reply:
> I suspect that new code should also work on Linux but haven't tested it (`make test` does not seem to run tests for patched functions, only `-finstrument-functions` and `-pg`, perhaps that should be fixed?).

Added in #1623